doc/manual/metaprogramming.rst

```diff
@@ -226,89 +322,112 @@ the important distinction between the way ``a`` and ``b`` are used:
 
-    for op = (:+, :*, :&, :|, :$)
-      eval(quote
-        ($op)(a,b,c) = ($op)(($op)(a,b),c)
-      end)
-    end
+    julia> make_expr2(:+, 1, 2)
+    :(2 + 4)
 
-In this manner, Julia acts as its own preprocessor, and allows code
-generation from inside the language. The above code could be written
-slightly more tersely using the ``:`` prefix quoting form::
+    julia> ex = make_expr2(:+, 1, Expr(:call, :*, 5, 8))
+    :(2 + 5 * 8)
 
-    for op = (:+, :*, :&, :|, :$)
-      eval(:(($op)(a,b,c) = ($op)(($op)(a,b),c)))
-    end
+    julia> eval(ex)
+    42
 
-This sort of in-language code generation, however, using the
-``eval(quote(...))`` pattern, is common enough that Julia comes with a
-macro to abbreviate this pattern::
 
-    for op = (:+, :*, :&, :|, :$)
-      @eval ($op)(a,b,c) = ($op)(($op)(a,b),c)
-    end
+Macros
+------
 
-The :obj:`@eval` macro rewrites this call to be precisely equivalent to the
-above longer versions. For longer blocks of generated code, the
-expression argument given to :obj:`@eval` can be a block::
+Macros are the analogue of functions for expression generation at
+compile time. Just as a function maps a tuple of argument values to a
+return value, a macro maps a tuple of argument *expressions* to a returned
+*expression*.
 
-    @eval begin
-      # multiple lines
-    end
+Why macros
+~~~~~~~~~~
 
-Interpolating into an unquoted expression is not supported and will
-cause a compile-time error:
+Here is an extraordinarily simple macro:
 
 .. doctest::
 
-    julia> $a + b
-    ERROR: unsupported or misplaced expression $
+    julia> macro sayhello(name)
+	       :( println("Hello, ", $name, "!") )
+	   end
 
-.. _man-macros:
+The "return value" of this macro is the *interpolated* expression; that is,
+with the value of the variable ``name`` passed directly as an argument to
+:func:`println`. We can verify this with the **extremely
+important** function :func:`macroexpand`:
 
-Macros
-------
+.. doctest::
 
-Macros are the analogue of functions for expression generation at
-compile time. Just as functions map a tuple of argument values to a 
-return value, macros map a tuple of argument *expressions* to a returned
-*expression*. They allow the programmer to arbitrarily transform the
-written code to a resulting expression, which then takes the place of
-the macro call in the final syntax tree. Macros are invoked with the
-following general syntax::
+    julia> ex = macroexpand( :(@sayhello("human")) )
+    :(println("Hello, ","human","!"))
+
+    julia> typeof(ex)
+    Expr
+
+:func:`macroexpand` takes a *quoted* macro and returns the resulting
+:obj:`Expr`\ession *after interpolation*.
+
+**Hold up.**
+
+We have already seen a function ``f(::Expr...) -> Expr`` in a previous
+section. In fact, :func:`macroexpand` is such a function itself;
+evaluating the result of :func:`macroexpand` will give the expected
+output:
+
+.. doctest::
+
+    julia> eval(macroexpand( quote @sayhello("human") end ))
```

Comment:
Amplifying my comment below, it's a little strange to have this here before the "Macro invocation" section and the introduction of `@`.